<a href="https://colab.research.google.com/github/roscibely/Sistemas-de-controle-Lab/blob/main/Controle_no_Espa%C3%A7o_de_Estados_Observadores_de_Estado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Controle no Espaço de Estados: Observadores de Estado

O observador de estados consiste em um mecanismo para estimação dos estados do sistema. É uma solução útil quando os estados reais da planta não estão accessíveis, situação muito comum na prática.

In [ ]:
!pip install control #Necessário para fazer o download da biblioteca control

## 1.  Bibliotecas necessárias






In [3]:
import numpy as np
import control as control
import matplotlib.pyplot as plt
from control.matlab import *

## 2. Definição do modelo no espaço de estado

Considere o seguinte sistema 

$$\dot{x}(t) = \left[\begin{array}{cc}-1 & 2 \\ 2 & -1 \end{array} \right] x(t)+ \left[\begin{array}{c}1 \\ 0 \end{array} \right] u(t)$$

$$y(t)= \left[\begin{array}{cc}0 & 1 \end{array} \right] x(t)$$

In [18]:
#######  Espaço de estado ########
A=np.matrix([[-1, 2],
             [2, -1]])
B=np.matrix([[1],
            [0]]) 
C=np.array([[0, 1]])
D=0

## 3. Observador

**Observabilidade**
  * Definição: O sistema $(A,B,C,D)$ é observável se para todo $x(0)$, o conhecimento da entrada $u(t)$ e da saída $y(t)$ em um tempo finito é suficiente para determinar $x(t)$;

  * *Teorema: O sistema $(A,B,C,D)$ é observável se e somente se o posto da matriz de observabilidade é igual a ordem do sistema;*

  A matriz de observabilidade é dada por: 
  $$O = \left[C\:\: CA \:\: CA^2 \: \ldots \: CA^{n-1}\right]^T $$

## 4. Calculando o ganho do observador de estado com o método de Ackermann

* Projete um observador de estado de modo a alocar os autovalores de $A-LC$ em $[-12,-12]$; 


In [39]:
L = acker(A.T, C.T, [-12,-12])
print("L=",L)

L= [[62.5 22. ]]


A função $acker$ realizar o calculo da matriz 
$$L^T= \left[\begin{array}{cccc}0 & 0 & \ldots & 1 \end{array} \right] \left[\begin{array}{cccc}C^T & A^TC^T & \ldots & (A^{n-1})^TC^T \end{array} \right] ^T \Delta(A)$$

* Considerando o sistema em questão com os pólos desejados $po=[-12, -12]$, o polinômio característico desejado é dado por: 
$$\Delta (s) = (s+12)(s+12) = s^2 +24s +144$$

Portanto, 
$$\Delta(A) = \left[\begin{array}{cc}125& 44\\ 44& 125 \end{array} \right]  $$

A matriz $L$ é dada por 

$$L^T= \left[\begin{array}{cc}0 &  1 \end{array} \right] \left[\begin{array}{cc}0.5 & 1 \\ 0.5 & 0 \end{array} \right] ^T \left[\begin{array}{cc}125& 44\\ 44& 125 \end{array} \right] = \left[\begin{array}{cc}62.5 &  22 \end{array} \right]^T$$